### PR TITLE
rpi5: add pinctrl and scmi pinctrl Pin control protocol support

### PR DESCRIPTION
**Pull request by GrygiriiS**
Based on the PR #14 
- added generic pinctrl interface
- added bcm2712 pinctrl driver
- added scmi pinctrl protocol

tested with https://github.com/GrygiriiS/linux/commits/gs_scmi_pinctrl_am_1/

Not tested under XEN as it depends on:
ARM TF #16
XEN https://github.com/xen-troops/xen/pull/209

### CODE REVIEW

**Comment by GrygiriiS** [comment]
CC: @lorc @oleksiimoisieiev

**Comment by GrygiriiS** [comment]
Ready for review.

File: drivers/scmi-msg/scmi_pinctrl.c [hunk 1]
```diff
@@ -0,0 +1,810 @@
...
+#include <errno.h>
+
+#include <drivers/pinctrl.h>
+#include <drivers/scmi-msg.h>
+#include <drivers/scmi.h>
+#include <lib/utils.h>
+#include <lib/utils_def.h>
+
+#include "common.h"
+#define PP(fmt, ...) \
```
Comment (oleksiimoisieiev) [inline drivers/scmi-msg/scmi_pinctrl.c]: This is leftover from the debug code

File: drivers/scmi-msg/scmi_pinctrl.c [hunk 2]
```diff
@@ -0,0 +1,810 @@
...
+static void scmi_pinctrl_report_attributes(struct scmi_msg *msg)
+{
+	uint16_t nr_functions = pinctrl_get_functions_count();
+	uint16_t nr_groups = pinctrl_get_groups_count();
+	unsigned nr_pins = pinctrl_get_pins_count();
+	struct scmi_pinctrl_proro_attrs_a2p {
+		int32_t status;
+		uint32_t attrs_low;
+		uint32_t attrs_high;
+	} return_values = {.status = SCMI_SUCCESS};
```
Comment (oleksiimoisieiev) [inline drivers/scmi-msg/scmi_pinctrl.c:172]: Please run checkpatch on commits. I've got a lot of warnings and even errors.

File: drivers/scmi-msg/scmi_pinctrl.c [hunk 3]
```diff
@@ -0,0 +1,810 @@
...
+#pragma weak pinctrl_get_function_name
+#pragma weak pinctrl_get_group_name
+#pragma weak pinctrl_get_group_pins
+#pragma weak pinctrl_get_function_groups
+#pragma weak pinctrl_request
+#pragma weak pinctrl_free
+
+uint16_t pinctrl_get_groups_count()
+{
+	return 0;
```
Comment (oleksiimoisieiev) [inline drivers/scmi-msg/scmi_pinctrl.c:41]: I think here and below should be -ENOTSUP

File: drivers/scmi-msg/scmi_pinctrl.c [hunk 4]
```diff
@@ -0,0 +1,810 @@
...
+static bool message_id_is_supported(unsigned int message_id);
+
+static void scmi_pinctrl_report_version(struct scmi_msg *msg)
+{
+	struct scmi_protocol_version_p2a return_values = {
+		.status = SCMI_SUCCESS,
+		.version = SCMI_PROTOCOL_VERSION_PINCTRL,
+	};
+
+	PP("scmi: Process pinctrl protocol_version\n");
```
Comment (oleksiimoisieiev) [inline drivers/scmi-msg/scmi_pinctrl.c]: Please replace this with VERBOSE

File: drivers/scmi-msg/scmi_pinctrl.c [hunk 5]
```diff
@@ -0,0 +1,810 @@
...
+
+	if (msg->in_size != 0U) {
+		PP("");
```
Comment (oleksiimoisieiev) [inline drivers/scmi-msg/scmi_pinctrl.c]: please remove

File: drivers/scmi-msg/scmi_pinctrl.c [hunk 6]
```diff
@@ -0,0 +1,810 @@
...
+	if (msg->in_size != 0U) {
+		scmi_status_response(msg, SCMI_PROTOCOL_ERROR);
+		return;
+	}
+
+	return_values.attrs_low =
+		FLD(GENMASK(31, 16), nr_groups) | FLD(GENMASK(15, 0), nr_pins);
+	return_values.attrs_high = FLD(GENMASK(15, 0), nr_functions);
+
+	INFO("scmi: pinctrl protocol_attrs group_max = %d, func_max = "
```
Comment (oleksiimoisieiev) [inline drivers/scmi-msg/scmi_pinctrl.c]: this probably needs to be verbose as well

File: plat/rpi/rpi5/platform.mk [hunk 7]
```diff
@@ -99,9 +99,16 @@ BL31_SOURCES		+= drivers/scmi-msg/base.c			\
 				drivers/scmi-msg/entry.c		\
 				drivers/scmi-msg/smt.c			\
 				drivers/scmi-msg/reset_domain.c		\
+				drivers/scmi-msg/scmi_pinctrl.c		\
+				drivers/pinctrl/pinctrl.c		\
 				plat/rpi/rpi5/scmi/scmi.c		\
 				plat/rpi/rpi5/scmi/scmi_reset.c		\
-				plat/rpi/rpi5/rpi5_svc_setup.c
+				plat/rpi/rpi5/rpi5_svc_setup.c \
+				plat/rpi/rpi5/bcm2712_pinctrl.c
```
Comment (oleksiimoisieiev) [inline plat/rpi/rpi5/platform.mk]: Probably it's better to split these changes between commts that are adding protocol and pinctrl driver? What do you tihnk?

**Review by oleksiimoisieiev — CHANGES_REQUESTED** [review]

File: drivers/scmi-msg/scmi_pinctrl.c (hunk 3 above)
Comment (GrygiriiS) [reply drivers/scmi-msg/scmi_pinctrl.c:41]: No get _count should return 0 as safe value.

File: plat/rpi/rpi5/platform.mk (hunk 7 above)
Comment (GrygiriiS) [reply plat/rpi/rpi5/platform.mk]: Not sure it make sense as pinctrl feature should be enabled all together.

File: drivers/scmi-msg/scmi_pinctrl.c (hunk 1 above)
Comment (GrygiriiS) [reply drivers/scmi-msg/scmi_pinctrl.c]: fixed

File: drivers/scmi-msg/scmi_pinctrl.c (hunk 2 above)
Comment (GrygiriiS) [reply drivers/scmi-msg/scmi_pinctrl.c:172]: done

File: drivers/scmi-msg/scmi_pinctrl.c (hunk 4 above)
Comment (GrygiriiS) [reply drivers/scmi-msg/scmi_pinctrl.c]: fixed

File: drivers/scmi-msg/scmi_pinctrl.c (hunk 5 above)
Comment (GrygiriiS) [reply drivers/scmi-msg/scmi_pinctrl.c]: fixed

File: drivers/scmi-msg/scmi_pinctrl.c (hunk 6 above)
Comment (GrygiriiS) [reply drivers/scmi-msg/scmi_pinctrl.c]: fixed

File: plat/rpi/rpi5/platform.mk (hunk 7 above)
Comment (GrygiriiS) [reply plat/rpi/rpi5/platform.mk]: updated

**Comment by GrygiriiS** [comment]
Optimized generic pinctrl interface

**Comment by oleksiimoisieiev** [comment]
LGTM `Reviewed-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>`

**Comment by GrygiriiS** [comment]
Closed as all work is now done in PR #18

**Pull request closed without merging**